### PR TITLE
Adding I2C functionality to the FX2 device. (Tested with a 24C256 EEPROM)

### PR DESCRIPTION
**Pull request by RacingTornado**
This commit adds I2C functionality to the eeprom. The same steps are applicable to this pull request  to make and build use the following steps:
To run this first clone the repository, then cd to the fx2lib directory and checkout this particular branch.
After this run 2 commands:
`git submodule init`
`git submodule update`
`make`
After this , navigate to the examples/i2c_bitbang directory and execute the following commands:
`make`
Now push the FX2 device inside the USB port(remove any udev rules you may have set up for Cypresse's VID and PID) and run
`make fx2_download`
After this open minicom 
`sudo minicom -H -w name_of_configuration`
You should see the data which is read from the EEPROM displayed on your console. Make sure the UART0 pins are connected to the appropriate GPIO's on the FX2. 

A couple of things to take care are that the I2C is in non-blocking mode. So often the same data is printed(that is previous data) if new data has not arrived. By using a logic analyzer the I2C can be confirmed to work. Otherwise we would need to busy wait in the loop which removes any benefit we may get from using queues.

### CODE REVIEW

File: examples/i2c_bitbang/i2c_main.c [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+
+void main() {
+   SETCPUFREQ(CLK_48M);
+   EA = 1;			// global interrupt enable
+   /********************************
+   I2C BLOCK BITBANG
+   *********************************/
+   //Called with number of retries
+   i2c_init(3);
```
Comment (mithro) [inline examples/i2c_bitbang/i2c_main.c:53]: 3?

File: examples/i2c_bitbang/i2c_main.c [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+   wr_addr = 0x03;
+   while (TRUE)
+    {
+
+        write_addr[0] = 0xa0;
+        write_addr[1] = 0x00;
+        write_addr[2] = wr_addr;
+        write_data[0] = 0x44;
+        //Address, data, address length and data length
+        I2CPutTX(&write_addr[0],&write_data[0],0x03,0x01);
```
Comment (mithro) [inline examples/i2c_bitbang/i2c_main.c:65]: Why `I2CPutTX` and `i2c_control()`? You are not being consistent again

File: examples/i2c_bitbang/i2c_main.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    mov _tx_i2c_buffer,a	//(2 cycles. Move the data back into accumulator for next ISR execution routine)
+    nop				//(1 cycle. Wait for 1 cycle before setting the bit high again.) 
+    sclh:
+    setb _PA6			//(2 cycles. Set the SCL high)
+    finish:
+    nop
+    __endasm;
+}
+
+/**************************
```
Comment (mithro) [inline examples/i2c_bitbang/i2c_main.c:121]: Why is it even in this file then? You know you can have multiple files.

File: examples/i2c_bitbang/i2c_main.c [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        }
+        else
+        {
+            wr_addr--;
+        }
+    }
+}
+
+void timer1_isr ()
+__interrupt TF1_ISR
```
Comment (mithro) [inline examples/i2c_bitbang/i2c_main.c:85]: Rather than checking the state in the ISR, set the ISR function to be the correct callback for a given state.

**Comment by mithro** [comment]
This needs a huge amount of work.
- You fail on consistency, formatting, etc.
- The queue structure you have created does a huge amount of work which seems pointless.
- You are confusing low level details with high level details.
- The I2C state machine is _really_ hard to follow.
- You don't deal with things like 8bit verse 16bit I2C addresses.

File: examples/i2c_bitbang/i2c_main.c (hunk 1 above)
Comment (RacingTornado) [inline examples/i2c_bitbang/i2c_main.c:53]: Retries 3 times.
